feat(forms): add `emitEvent` option for AbstractControl-based class methods

MikeJerred opened the pull request:
This commit adds an optional argument to the methods addControl
and removeControl in the FormGroup class. This argument can be
used to prevent an event from being emitted.

PR Close #29662

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #29662 


## What is the new behavior?
An argument can now be specified to prevent an event from being emitted.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

MikeJerred commented:
I don't know why the build is failing, it was working in PR #29664 which has exactly the same code changes

AndrewKushnir commented:
Hi @MikeJerred, sorry we didn't have a chance to review/reply earlier.

As I mentioned in https://github.com/angular/angular/issues/20439#issuecomment-763976919, we'd like to consider adding `emitEvent` config to the `FormGroup` and `FormArray` classes. This PR covers most of the functions already (except the `FormGroup.setControl` one) and has good test coverage, so I'm wondering if you might be interested in rebasing this PR and adding the `emitEvent` support to the `FormGroup.setControl` method as well? If you don't have time to look into that now, I can help with rebasing and extending `FormGroup.setControl` API.

Thank you.

MikeJerred commented:
@AndrewKushnir I have rebased this PR and added the changes to `FormGroup.setControl`. The tests are failing but as far as I can tell the failure is not related to my changes - perhaps you could help with debugging that?

AndrewKushnir commented:
Hi @MikeJerred, thanks for updating the PR!

The CI failure was a flake, restarting that CI job resolved the issue.

I will review the changes within the next couple days and get back with comments/feedback.

Thank you.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1450,22 +1450,36 @@ export class FormGroup extends AbstractControl {
    *
    * @param name The control name to add to the collection
    * @param control Provides the control for the given name
+   * @param opts Configuration options determine how the control propagates changes and emits
+   * events after the control is added.
```
Since the `onlySelf` is not included into the options (which looks correct, if having `onlySelf` support would be needed in the future - it'd be a separate PR), we should update the docs to reflect that:
```suggestion
   * @param opts Allows to specify whether this FormGroup instance should emit events after a new control is added.
```
I believe docs for other functions should be rephrased a bit as well.

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -64,6 +64,91 @@ describe('FormArray', () => {
     });
+
+    it('should not emit event when pushing if emitEvent is false', fakeAsync(() => {
+         a.valueChanges.subscribe((value) => {
+           throw 'Should not happen';
+         });
```
I've looked at the code of this spec file and it seems like it's more common to log events and check the log array output:
```suggestion
         a.valueChanges.subscribe(() => logger.push('valueChanges'));
```
It's pretty minor, but would be great to have tests in a bit more consistent state.

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -64,6 +64,91 @@ describe('FormArray', () => {
+         a.valueChanges.subscribe((value) => {
+           throw 'Should not happen';
+         });
+         a.statusChanges.subscribe((value) => {
+           throw 'Should not happen';
+         });
```
I think it'd be great to have cases where `FormGroup` and `FormArray` have parent `FormGroup`s, so that we can make sure that `emitEvents` option controls not only a current level, but also a parent group/array.

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -64,6 +64,91 @@ describe('FormArray', () => {
+         });
+
+         a.push(c1, {emitEvent: false});
+         tick();
```
It looks like `tick()` call and `fakeAsync(...)` wrapper are not needed since these APIs are synchronous. They are used in other tests where there is an async validator defined (thus we need to call tick to emulate the async behavior and verify the results).

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -64,6 +64,91 @@ describe('FormArray', () => {
       expect(a.controls).toEqual([c1, c2, c3]);
     });
+
+    it('should not emit event when pushing if emitEvent is false', fakeAsync(() => {
```
I'd propose to format test title like this:
```suggestion
    it('should not emit events when calling `FormArray.push` with `emitEvent: false`', fakeAsync(() => {
```
So that it's easier to understand what the test does just by looking at the spec title.

(similar in other test cases too)

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -148,6 +148,45 @@ describe('FormGroup', () => {
+
+    it('should not emit events when control is removed and emitEvent is false', fakeAsync(() => {
+         const g = new FormGroup(
+             {'one': new FormControl('1'), 'two': new FormControl('2', Validators.minLength(10))});
```
For completeness, I think it'd be great to have some test(s) that would verify that when `emitEvents` is set to `false`, there are no events emitted in case adding a `FormControl` changes `FormGroup` status, for example:

```
// Validator function which would return error in case there is a control with 'invalidCtrl' is added. 
const validatorFn = (value) => !!value.invalidCtrl ? {invalidCtrl: true} : null /* success */;
const asyncValidatorFn = (value) => of(validatorFn(value));

const g = new FormGroup({}, validatorFn, asyncValidatorFn);
g.addControl('one', new FormControl(''), {emitEvent: false});
// expect g.valid to be `true`

g.addControl('invalidCtrl', new FormControl(''), {emitEvent: false});
// expect g.valid to be `false`, but there should be no events emitted
```

What do you think?

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -1864,6 +1903,14 @@ describe('FormGroup', () => {
+    it('should not emit event if emitEvent is false', () => {
+      const logger: string[] = [];
+      const c2 = new FormControl('new!');
+      g.valueChanges.subscribe(() => logger.push('change!'));
```
```suggestion
      g.statusChanges.subscribe(() => logger.push('statusChanges'));
      g.valueChanges.subscribe(() => logger.push('valueChanges'));
```

AndrewKushnir reviewed (changes requested):
@MikeJerred thanks for updating the PR! The changes look good, I've just left a few comments related to docs and tests. Please let me know if you have any questions and/or need any help. Thank you.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1450,22 +1450,36 @@ export class FormGroup extends AbstractControl {
+   * This method also updates the value and validity of the control.
+   *
    * @param name The control name to remove from the collection
+   * @param opts Allows to specify whether this FormGroup instance should emit events after a new
+   *     control is added.
```
```suggestion
   * @param opts Allows to specify whether this FormGroup instance should emit events after a
   *     control is removed.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1474,12 +1488,17 @@ export class FormGroup extends AbstractControl {
    *
    * @param name The control name to replace in the collection
    * @param control Provides the control for the given name
+   * @param opts Allows to specify whether this FormGroup instance should emit events after a new
+   *     control is added.
```
```suggestion
   * @param opts Allows to specify whether this FormGroup instance should emit events after an existing
   *     control is replaced.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1474,12 +1488,17 @@ export class FormGroup extends AbstractControl {
+   *     control is added.
+   * * `emitEvent`: When true or not supplied (the default), both the `statusChanges` and
+   * `valueChanges` observables emit events with the latest status and value when the control is
+   * removed. When false, no events are emitted.
```
```suggestion
   * replaced with a new one. When false, no events are emitted.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1876,11 +1895,16 @@ export class FormArray extends AbstractControl {
    * Insert a new `AbstractControl` at the end of the array.
    *
    * @param control Form control to be inserted
+   * @param opts Allows to specify whether this FormGroup instance should emit events after a new
+   *     control is added.
```
```suggestion
   * @param opts Allows to specify whether this FormArray instance should emit events after a new
   *     control is added.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1889,32 +1913,47 @@ export class FormArray extends AbstractControl {
    *
    * @param index Index in the array to insert the control
    * @param control Form control to be inserted
+   * @param opts Allows to specify whether this FormGroup instance should emit events after a new
+   *     control is added.
```
```suggestion
   * @param opts Allows to specify whether this FormArray instance should emit events after a new
   *     control is inserted.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1889,32 +1913,47 @@ export class FormArray extends AbstractControl {
    * Remove the control at the given `index` in the array.
    *
    * @param index Index in the array to remove the control
+   * @param opts Allows to specify whether this FormGroup instance should emit events after a new
+   *     control is added.
+   * * `emitEvent`: When true or not supplied (the default), both the `statusChanges` and
+   * `valueChanges` observables emit events with the latest status and value when the control is
+   * inserted. When false, no events are emitted.
```
```suggestion
   * @param opts Allows to specify whether this FormArray instance should emit events after a
   *     control is removed.
   * * `emitEvent`: When true or not supplied (the default), both the `statusChanges` and
   * `valueChanges` observables emit events with the latest status and value when the control is
   * removed. When false, no events are emitted.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1889,32 +1913,47 @@ export class FormArray extends AbstractControl {
    *
    * @param index Index in the array to replace the control
    * @param control The `AbstractControl` control to replace the existing control
+   * @param opts Allows to specify whether this FormGroup instance should emit events after a new
+   *     control is added.
```
```suggestion
   * @param opts Allows to specify whether this FormArray instance should emit events after an existing
   *     control is replaced with a new one.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1889,32 +1913,47 @@ export class FormArray extends AbstractControl {
+   *     control is added.
+   * * `emitEvent`: When true or not supplied (the default), both the `statusChanges` and
+   * `valueChanges` observables emit events with the latest status and value when the control is
+   * inserted. When false, no events are emitted.
```
```suggestion
   * replaced with a new one. When false, no events are emitted.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -2095,6 +2134,12 @@ export class FormArray extends AbstractControl {
   /**
    * Remove all controls in the `FormArray`.
    *
+   * @param opts Allows to specify whether this FormGroup instance should emit events after a new
+   *     control is added.
```
```suggestion
   * @param opts Allows to specify whether this FormArray instance should emit events after all
   *     controls are removed.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -2095,6 +2134,12 @@ export class FormArray extends AbstractControl {
+   * @param opts Allows to specify whether this FormGroup instance should emit events after a new
+   *     control is added.
+   * * `emitEvent`: When true or not supplied (the default), both the `statusChanges` and
+   * `valueChanges` observables emit events with the latest status and value when the control is
+   * cleared. When false, no events are emitted.
```
```suggestion
   * `valueChanges` observables emit events with the latest status and value when all controls
   * in this FormArray instance are removed. When false, no events are emitted.
```

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -18,12 +18,14 @@ describe('FormArray', () => {
   describe('adding/removing', () => {
     let a: FormArray;
     let c1: FormControl, c2: FormControl, c3: FormControl;
+    let logger: any[];
```
```suggestion
    let logger: string[];
```

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -148,6 +154,56 @@ describe('FormGroup', () => {
+         g.valueChanges.subscribe(() => logger.push('value change'));
+         g.statusChanges.subscribe(() => logger.push('status change'));
+
+         g.addControl('two', new FormControl('2', Validators.minLength(10)), {emitEvent: false});
```
I think having a `minLength` validator in this test is not required, so we can remove it:
```suggestion
         g.addControl('two', new FormControl('2'), {emitEvent: false});
```

MikeJerred replied:
Ah OK, I had added that to check that `statusChanges` did not emit when the status changed (in this case `VALID` -> `INVALID`), but it seems that `statusChanges` emits every time it recalculates, not just when there is a change, so you are right and this validator is unnecessary.

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -148,6 +154,56 @@ describe('FormGroup', () => {
+         expect(logger).toEqual([]);
+       });
+
+    it('should not emit status change events when `FormGroup.addControl` is called with `emitEvent: false`',
```
I think it'd be great to have similar test for `FormArray` as well.

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -307,6 +358,16 @@ describe('FormGroup', () => {
         expect(logger).toEqual(['control1', 'control2', 'group', 'form']);
       });
+
+      it('should not emit events on the parent when called with `emitEvent: false`', () => {
```
It'd be great to have a similar test for `FormArray` too.

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -307,6 +358,16 @@ describe('FormGroup', () => {
+        form.statusChanges.subscribe(() => logger.push('form status change'));
+        g.statusChanges.subscribe(() => logger.push('group status change'));
+
+        g.addControl('three', new FormControl(5, Validators.min(10)), {emitEvent: false});
```
It looks like validator is not needed in this case too:
```suggestion
        g.addControl('three', new FormControl(5), {emitEvent: false});
```

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -1864,6 +1914,15 @@ describe('FormGroup', () => {
       expect(logger).toEqual(['change!']);
     });
+
+    it('should not emit event called with `emitEvent: false`', () => {
```
```suggestion
    it('should not emit event called when `FormGroup.setControl` with `emitEvent: false`', () => {
```

AndrewKushnir reviewed (commented):
Thanks for additional changes @MikeJerred 👍 

I've left a few more comments mostly around docs to rephrase them a bit to be aligned with the function itself.

Just as a thought for a future refactoring (later in a separate PR), it'd be great to make description of the `opts` argument more generic (do not name the action specifically, but just refer to it as the "current action"), so that it's easier to maintain. The docs may look like:

```
   * @param opts Allows to specify whether events should be emitted once the current 
   * operation is completed.
   * * `emitEvent`: When true or not supplied (the default), both `statusChanges` and 
   * `valueChanges` events are emitted with the latest status and value. When false, no 
   * events are emitted.
```
We'd probably need to revisit exiting functions as well and update `opts` argument docs in a similar way. But this is definitely outside of the scope of this PR (I'll create a ticket once this PR lands).

Thank you.

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -64,6 +66,98 @@ describe('FormArray', () => {
+       () => {
+         const validatorFn = (value: any) => value.controls.length > 0 ? {controls: true} : null;
+         const asyncValidatorFn = (value: any) => of(validatorFn(value));
+         a = new FormArray([], validatorFn, asyncValidatorFn);
```
Minor note: I'd propose creating a new const here to avoid the need to search for the current `a` definition.
```suggestion
         const arr = new FormArray([], validatorFn, asyncValidatorFn);
```

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -64,6 +66,98 @@ describe('FormArray', () => {
+
+    it('should not emit status change events when `FormArray.push` is called with `emitEvent: false`',
+       () => {
+         const validatorFn = (value: any) => value.controls.length > 0 ? {controls: true} : null;
```
Could you plz add a comment on why we've added validators here?
```suggestion
         // Adding validators to make sure there are no status change event submitted when form becomes invalid.
         const validatorFn = (value: any) => value.controls.length > 0 ? {controls: true} : null;
```

AndrewKushnir reviewed (approved):
Thanks for the updates @MikeJerred, the changes look great, I've just added a couple final comments.

Could you please also squash commit and update commit message to include the list of methods that were updated? Proposed commit message:

```
feat(forms): add `emitEvent` option for AbstractControl-based class methods

This commit adds the `emitEvent` option to the following FromArray and FormGroup methods:

* FormGroup.addControl
* FormGroup.removeControl
* FormGroup.setControl
* FormArray.push
* FormArray.insert
* FormArray.removeAt
* FormArray.setControl
* FormArray.clear

This option can be used to prevent an event from being emitted when adding or removing controls.

Closes #29662.
```

I've also started [tests for this change in Google's codebase](http://test/OCL:355534001:BASE:355534007:1612409245205:6b32e1a0) (internal-only link) and will share results once I have them.

atscott commented on goldens/public-api/forms/forms.d.ts:
```diff
@@ -172,20 +172,30 @@ export declare class FormArray extends AbstractControl {
         emitEvent?: boolean;
     }): void;
-    setControl(index: number, control: AbstractControl): void;
+    setControl(index: number, control: AbstractControl, opts?: {
```
naming nit: The parameter name is `options` for the existing functions that allow `emitEvent` configuration. Let's keep them consistent.

MikeJerred replied:
`FormControl` uses `opts` for this, shall I change that to `options` too?

AndrewKushnir replied:
Looking at the current version of the [forms.d.ts file](https://github.com/angular/angular/blob/master/goldens/public-api/forms/forms.d.ts), the `options` is more common that `opts` (14 vs 10 usages) and the FormGroup and FormArray methods mostly use full form (i.e. `options`). I think it makes sense to use full form in this PR as well (as Andrew suggested) and may be later (as a separate refactoring), we can rename all other `opts` to `options`.

atscott reviewed (approved):
Small nit for the parameter name but otherwise LGTM

reviewed-for: public-api

AndrewKushnir commented:
Quick update after running tests in Google's codebase: we've came across some cases where `FormGroup` and `FormArray` classes are extended and some methods have slightly different signature which become incompatible with the changes from this PR. That's an indication that we should consider this PR a breaking change (I've updated labels on this PR accordingly). I will run more broad set of tests in Google's database to find all instances of custom controls that extend `FormGroup` and `FormArray` and will provide an update as soon as I get results. Thank you.

rafaelss95 commented on packages/forms/src/model.ts:
```diff
@@ -1450,22 +1450,36 @@ export class FormGroup extends AbstractControl {
-    delete (this.controls[name]);
-    this.updateValueAndValidity();
+  removeControl(name: string, options: {emitEvent?: boolean} = {}): void {
+    if (this.controls[name]) {
+      this.controls[name]._registerOnCollectionChange(() => {});
+      delete (this.controls[name]);
+    }
```
Excuse me, but is this change made on purpose? I'm asking it because before, the `delete` operation occurred even though `this.controls [name]` corresponds to a _falsy_ value and now it is only executed when `this.controls [name]` matches a _truthy_ value.

AndrewKushnir replied:
Thanks for the comment @rafaelss95. The thing is that `this.controls` is typed as `{[key: string]: AbstractControl}`, so we only expect instances of AbstractControl-based classes to be present here as values. Please let me know if we are missing something and you can come up with a case where moving the `delete` operation into the `if` block might cause issues?

petebacondarwin commented on packages/forms/src/model.ts:
```diff
@@ -1450,22 +1450,36 @@ export class FormGroup extends AbstractControl {
    *
    * @param name The control name to add to the collection
    * @param control Provides the control for the given name
+   * @param options Allows to specify whether this FormGroup instance should emit events after a new
```
Docs NIT: is `Allows to` necessary?

```suggestion
   * @param options Specify whether this FormGroup instance should emit events after a new
```

petebacondarwin commented on packages/forms/src/model.ts:
```diff
@@ -1450,22 +1450,36 @@ export class FormGroup extends AbstractControl {
-    delete (this.controls[name]);
-    this.updateValueAndValidity();
+  removeControl(name: string, options: {emitEvent?: boolean} = {}): void {
+    if (this.controls[name]) {
```
Having got bitten by this recently... should we be more careful to use something like `Object.hasOwnProperty(this.controls, name)` or . I know it is unlikely but what if there was a control called `toString` etc?

AndrewKushnir replied:
I think we could just revert this change as it's completely unrelated to this PR and we can do it later as a separate cleanup/refactoring PR.

petebacondarwin commented on packages/forms/src/model.ts:
````diff
@@ -2122,11 +2167,11 @@ export class FormArray extends AbstractControl {
    * ```
    */
-  clear(): void {
+  clear(options: {emitEvent?: boolean} = {}): void {
````
Since `updateValueAndValidity()`  can also take `undefined` then there is no real need to provide a default value in this and all the other methods that just pass the `options` parameter through.

petebacondarwin commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -64,6 +66,100 @@ describe('FormArray', () => {
+
+    it('should not emit status change events when `FormArray.push` is called with `emitEvent: false`',
+       () => {
+         // Adding validators to make sure there are no status change event submitted when form
```
```suggestion
         // Adding validators to make sure there are no status change events submitted when form
```

petebacondarwin reviewed (changes requested):
Nice work @MikeJerred - loving the tests.

I wonder... given that the `options` object is always just passed through, then at runtime we could also support `onlySelf` too. Does that make sense in these scenarios? Should we expose that too?

Or perhaps more pertinent, if it definitely doesn't make sense and shouldn't appear in the `options` object, then perhaps we should be excluding that property... E.g.

```ts
insert(index: number, control: AbstractControl, options: {emitEvent?: boolean} = {}): void {
  this.controls.splice(index, 0, control);
  this._registerControl(control);
  this.updateValueAndValidity({emitEvent: options.emitEvent});  
}
```

So as ensure a caller doesn't accidentally set the `onlySelf` property by calling with:

```ts
  const options = { onlySelf: true, emitEvent: false };
  container.insert(2, ctrl, options);
```

jessicajaniuk commented on packages/forms/src/model.ts:
```diff
@@ -1450,22 +1450,36 @@ export class FormGroup extends AbstractControl {
    *
    * @param name The control name to add to the collection
    * @param control Provides the control for the given name
+   * @param options Allows to specify whether this FormGroup instance should emit events after a new
```
As an alternative suggestion here:
```suggestion
   * @param options Allows the ability to specify whether this FormGroup instance should emit events after a new
```

jessicajaniuk replied:
Or possibly:
```suggestion
   * @param options Specifies whether this FormGroup instance should emit events after a new
```

jessicajaniuk reviewed (approved):
Left a minor suggestion, but otherwise LGTM 🍪 

reviewed-for: public-api

AndrewKushnir commented:
@petebacondarwin thanks for the review, great points! 👍 

> I wonder... given that the options object is always just passed through, then at runtime we could also support onlySelf too. Does that make sense in these scenarios? Should we expose that too?
>
> Or perhaps more pertinent, if it definitely doesn't make sense and shouldn't appear in the options object, then perhaps we should be excluding that property... E.g.

Just wanted to leave a quick comment on the `onlySelf` config option (I was also thinking about it): since updated functions perform structural changes to control tree (i.e. adding/removing controls/collections), exposing `onlySelf` option (effectively allowing "hiding" such changes from parent controls) would require more research to make sure it's safe to do. With `emitEvents` it's pretty safe since the `valueChanges` and `statusChanges` are mostly used "externally" and that option just controls whether events should be broadcasted or not.

Based on the feedback in linked tickets, the `emitEvent` is the only option that was missing/required in use-cases, so I think we should just expose `emitEvents` and keep `onlySelf` for the next iteration (if we have enough evidence that it's actually needed).

+1 to make sure that we do not just proxy `options` to `updateValueAndValidity` function call allowing `onlySelf` to be passed in as well. Great catch! 👍

mary-poppins commented:
You can preview 21dad98 at https://pr31031-21dad98.ngbuilds.io/.
You can preview 2e305bd at https://pr31031-2e305bd.ngbuilds.io/.
You can preview b0ea8b2 at https://pr31031-b0ea8b2.ngbuilds.io/.
You can preview 07cce2a at https://pr31031-07cce2a.ngbuilds.io/.
You can preview 58dc91b at https://pr31031-58dc91b.ngbuilds.io/.
You can preview af4753f at https://pr31031-af4753f.ngbuilds.io/.
You can preview aa533c1 at https://pr31031-aa533c1.ngbuilds.io/.
You can preview 5a24fc6 at https://pr31031-5a24fc6.ngbuilds.io/.
You can preview 6072db2 at https://pr31031-6072db2.ngbuilds.io/.
You can preview 7218be5 at https://pr31031-7218be5.ngbuilds.io/.
You can preview e02d862 at https://pr31031-e02d862.ngbuilds.io/.

AndrewKushnir commented:
@MikeJerred thanks for applying the changes.

I noticed that you've changed the implementation for `options` from using default value to making the second argument optional. I think these options are close, but the default value approach is more consistent with existing APIs. Could you please update it?

Also I want to share the next steps:

- [ ] since this PR changes public API surface, more people from the `public-api` group would review the PR and provide feedback
- [ ] address the feedback from reviewers
- [ ] run tests in Google's codebase to make sure there are no breakages (we'll take care of that)
- [ ] add "BREAKING CHANGES" section to the commit message and describe that custom classes that extend AbstractControl-based ones and override existing methods might need some updates

Thank you.

petebacondarwin reviewed (approved):
Great. Thanks @MikeJerred

petebacondarwin reviewed (approved):
Reviewed-for: public-api

MikeJerred committed:
feat(forms): add `emitEvent` option for AbstractControl-based class methods

This commit adds the `emitEvent` option to the following FormArray and FormGroup methods:

* FormGroup.addControl
* FormGroup.removeControl
* FormGroup.setControl
* FormArray.push
* FormArray.insert
* FormArray.removeAt
* FormArray.setControl
* FormArray.clear

This option can be used to prevent an event from being emitted when adding or removing controls.

BREAKING CHANGE:

The `emitEvent` option was added to the following `FormArray` and `FormGroup` methods:

* FormGroup.addControl
* FormGroup.removeControl
* FormGroup.setControl
* FormArray.push
* FormArray.insert
* FormArray.removeAt
* FormArray.setControl
* FormArray.clear

If your app has custom classes that extend `FormArray` or `FormGroup` classes and override the 
above-mentioned methods, you may need to update your implementation to take the new options into
account and make sure that overrides are compatible from a types perspective.

Closes #29662.

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:357109561:BASE:357109577:1613101271368:e0c49fd9).

mary-poppins commented:
You can preview 21dad98 at https://pr31031-21dad98.ngbuilds.io/.
You can preview 2e305bd at https://pr31031-2e305bd.ngbuilds.io/.
You can preview b0ea8b2 at https://pr31031-b0ea8b2.ngbuilds.io/.
You can preview 07cce2a at https://pr31031-07cce2a.ngbuilds.io/.
You can preview 58dc91b at https://pr31031-58dc91b.ngbuilds.io/.
You can preview af4753f at https://pr31031-af4753f.ngbuilds.io/.
You can preview aa533c1 at https://pr31031-aa533c1.ngbuilds.io/.
You can preview 5a24fc6 at https://pr31031-5a24fc6.ngbuilds.io/.
You can preview 6072db2 at https://pr31031-6072db2.ngbuilds.io/.
You can preview 7218be5 at https://pr31031-7218be5.ngbuilds.io/.
You can preview e02d862 at https://pr31031-e02d862.ngbuilds.io/.
You can preview 473821f at https://pr31031-473821f.ngbuilds.io/.

AndrewKushnir commented:
@MikeJerred quick update: tests in Google's codebase went well (after minor updates to classes that extend FormGroup and FormArray). We are very close to merging this change and the last step would be to squash all commits into one and add the "BREAKING CHANGE" note to the commit body, so the commit message may look like this:

```
feat(forms): add `emitEvent` option for AbstractControl-based class methods

This commit adds the `emitEvent` option to the following FromArray and FormGroup methods:

* FormGroup.addControl
* FormGroup.removeControl
* FormGroup.setControl
* FormArray.push
* FormArray.insert
* FormArray.removeAt
* FormArray.setControl
* FormArray.clear

This option can be used to prevent an event from being emitted when adding or removing controls.

BREAKING CHANGE:

The `emitEvent` option was added to the following `FromArray` and `FormGroup` methods:

* FormGroup.addControl
* FormGroup.removeControl
* FormGroup.setControl
* FormArray.push
* FormArray.insert
* FormArray.removeAt
* FormArray.setControl
* FormArray.clear

If your app has custom classes that extend `FromArray` or `FormGroup` classes and override the 
above-mentioned methods, you may need to update your implementation to take new options into
account and make sure that overrides are compatible from types perspective.

Closes #29662.
```

Please let us know once it's ready, we'll do final checks and add this PR to the merge queue.

Thank you.

MikeJerred commented:
@AndrewKushnir I have squashed the commits into one and adjusted the message.

rafaelss95 commented:
@MikeJerred there are some typos in your commit: `FromArray` should be `FormArray`.

AndrewKushnir commented:
FYI, started a [new presubmit in Google's codebase](https://test.corp.google.com/OCL:357628881:BASE:357628982:1613440663102:7b0a449a) (internal-only link) to verify that there there are no places that might be broken by this change. Will keep this thread updated.

AndrewKushnir commented:
FYI all affected tests in Google's codebase are successful, so I'm adding this PR to the merge queue.

@MikeJerred thanks for working on this PR and addressing the feedback! The change will be merged into the master branch and should be available as a part of the next major release (v12) in a few weeks.
